sdktests/server_side_migrations.go

```diff
@@ -0,0 +1,370 @@
+		{Operation: ldmigration.Read, Stage: ldmigration.Complete, ValuesMatchers: onlyNew},
+
+		// Write operations
+		{Operation: ldmigration.Write, Stage: ldmigration.Off, ValuesMatchers: onlyOld},
+		{Operation: ldmigration.Write, Stage: ldmigration.DualWrite, ValuesMatchers: both},
+		{Operation: ldmigration.Write, Stage: ldmigration.Shadow, ValuesMatchers: both},
+		{Operation: ldmigration.Write, Stage: ldmigration.Live, ValuesMatchers: both},
+		{Operation: ldmigration.Write, Stage: ldmigration.RampDown, ValuesMatchers: both},
+		{Operation: ldmigration.Write, Stage: ldmigration.Complete, ValuesMatchers: onlyNew},
+	}
+
+	for _, testParam := range testParams {
+		t.Run(fmt.Sprintf("%s latency for %s", testParam.Operation, testParam.Stage), func(t *ldtest.T) {
+			client, events := createClient(t, int(testParam.Stage))
+
+			callback := func(w http.ResponseWriter, req *http.Request) {
+				time.Sleep(10 * time.Millisecond)
+				w.WriteHeader(http.StatusOK)
+				w.Write([]byte("result"))
+			}
+
+			service := mockld.NewMigrationCallbackService(requireContext(t).harness, t.DebugLogger(), callback, callback)
+			t.Defer(service.Close)
+
+			context := ldcontext.New("key")
+
+			params := servicedef.MigrationOperationParams{
+				Key:                "migration-key",
+				Context:            context,
+				DefaultStage:       ldmigration.DualWrite,
+				ReadExecutionOrder: ldmigration.Concurrently,
+				OldEndpoint:        service.OldEndpoint().BaseURL(),
+				NewEndpoint:        service.NewEndpoint().BaseURL(),
+				Operation:          testParam.Operation,
+				TrackLatency:       true,
+			}
+
+			_ = client.MigrationOperation(t, params)
+			client.FlushEvents(t)
+
+			opEventMatchers := []m.Matcher{
+				m.JSONOptProperty("samplingRatio").Should(m.BeNil()),
+				m.JSONProperty("operation").Should(m.Equal(testParam.Operation.String())),
+				m.JSONProperty("evaluation").Should(
+					m.AllOf(
+						m.JSONProperty("key").Should(m.Equal("migration-key")),
+						m.JSONProperty("default").Should(m.Equal("dualwrite")),
+						m.JSONProperty("value").Should(m.Equal(testParam.Stage.String())),
+						m.JSONProperty("variation").Should(m.Equal(int(testParam.Stage))),
+						m.JSONProperty("reason").Should(
+							m.JSONProperty("kind").Should(m.Equal("FALLTHROUGH")),
+						),
+					),
+				),
+				m.JSONProperty("measurements").Should(
+					m.ItemsInAnyOrder(
+						m.AllOf(
+							m.JSONProperty("key").Should(m.Equal("latency_ms")),
+							m.JSONProperty("values").Should(
+								m.AllOf(
+									testParam.ValuesMatchers...,
+								),
+							),
+						),
+					),
+				),
+			}
+
+			payload := events.ExpectAnalyticsEvents(t, defaultEventTimeout)
+			m.In(t).Assert(payload, m.ItemsInAnyOrder(
+				IsIndexEventForContext(context),
+				IsSummaryEvent(),
+				IsValidMigrationOpEventWithConditions(
+					context,
+					opEventMatchers...,
+				),
+			))
+		})
+	}
+}
+
+func runTrackErrorsTests(t *ldtest.T) {
```

Comment:
Because authoritative write failures should halt execution, testing for error tracking has many more cases than latency tracking.